packages/modules/order/src/utils/actions/item-update.ts

```diff
@@ -0,0 +1,69 @@
+import { ChangeActionType, MathBN, MedusaError } from "@medusajs/utils"
+import { OrderChangeProcessing } from "../calculate-order-change"
+import { setActionReference } from "../set-action-reference"
+
+OrderChangeProcessing.registerActionType(ChangeActionType.ITEM_UPDATE, {
+  operation({ action, currentOrder, options }) {
+    const existingIndex = currentOrder.items.findIndex(
+      (item) => item.id === action.details.reference_id
+    )
+
+    const existing = currentOrder.items[existingIndex]
+
+    existing.detail.quantity ??= 0
+
+    const quantityDiff = MathBN.sub(
+      action.details.quantity,
+      existing.detail.quantity
+    )
+
+    existing.quantity = action.details.quantity
+    existing.detail.quantity = action.details.quantity
+
+    setActionReference(existing, action, options)
+
+    if (MathBN.lte(existing.quantity, 0)) {
+      currentOrder.items.splice(existingIndex, 1)
+    }
+
+    return MathBN.mult(existing.unit_price, quantityDiff)
+  },
+  validate({ action, currentOrder }) {
+    const refId = action.details?.reference_id
+    if (refId == null) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        "Reference ID is required."
+      )
+    }
+
+    const existing = currentOrder.items.find((item) => item.id === refId)
+    if (!existing) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Item ID "${refId}" not found.`
+      )
+    }
+
+    if (action.details?.quantity == null) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Quantity of item ${refId} is required.`
+      )
+    }
+
+    action.details.quantity ??= 0
+
+    const lower = MathBN.lt(
+      action.details.quantity,
+      existing.detail?.fulfilled_quantity
+    )
+
+    if (lower) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Item ${refId} has already been fulfilled and quantity cannot be lower than ${existing.detail?.fulfilled_quantity}.`
+      )
+    }
```

Comment:
I am not sure I understand this check + error message, but might just be missing some context.

What we want to ensure is that you can only update the unfulfilled quantity of an item right? 

So there are a couple of different cases to solve for:

**Item One**
- quantity: 3
- fulfilled_quantity: 0
You can update the quantity to anything, even 0 (i.e. remove it). 

**Item Two**
- quantity: 3
- fulfilled_quantity: 3
You can increase the quantity by as much as you want, but you cannot decrease the quantity because it is fully fulfilled.

**Item Three**
- quantity: 3
- fulfilled_quantity: 2
You can increase the quantity by as much as you want, but only decrease the quantity by 1. The "decreasable" quantity is `quantity - fulfilled_quantity`. 

It is not clear to me if we are solving for all three cases here. Can you confirm or deny this?

Reply:
That is it. All these 3 examples are true.
This action "ITEM_UPDATE" is to set an arbitrary number in the quantity, different from "ITEM_ADD" and "ITEM_REMOVE".

So, we need to compare if the final quantity set is equal or bigger than what was fulfilled. Otherwise, we throw the error.

Reply:
Okay cool, sounds good 👍